providers/implementations/encode_decode/ml_dsa_codecs.h

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2025 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef PROV_ML_DSA_CODECS_H
+# define PROV_ML_DSA_CODECS_H
+# pragma once
+
+# include <openssl/e_os2.h>
+# include "crypto/ml_dsa.h"
+# include "prov/provider_ctx.h"
+
+ /*-
+  * The DER ASN.1 encoding of ML-KEM (and ML-DSA) public keys prepends 22 bytes
+  * to the encoded public key:
+  *
+  * - 4 byte outer sequence tag and length
+  * -  2 byte algorithm sequence tag and length
+  * -    2 byte algorithm OID tag and length
+  * -      9 byte algorithm OID (from NIST CSOR OID arc)
+  * -  4 byte bit string tag and length
+  * -    1 bitstring lead byte
+  */
+# define ML_DSA_SPKI_OVERHEAD   22
+typedef struct {
+    const uint8_t asn1_prefix[ML_DSA_SPKI_OVERHEAD];
+} ML_DSA_SPKI_FMT;
+
+/*-
+* For each parameter set we support a few PKCS#8 input formats, three
+ * corresponding to the "either or both" variants of:
+ *
+ *  ML-DSA-PrivateKey ::= CHOICE {
+ *    seed [0] IMPLICIT OCTET STRING SIZE (32),
+ *    expandedKey OCTET STRING SIZE (2560 | 4032 | 4896)
+ *    both SEQUENCE {
+ *      seed OCTET STRING SIZE (32),
+ *      expandedKey OCTET STRING SIZE (2560 | 4032 | 4896) } }
+ *
+ * one more for a historical OQS encoding:
+ *
+ * - OQS private + public key: OCTET STRING
+ *   (The public key is ignored, just as with PKCS#8 v2.)
+ *
+ * and two more that are the minimal IETF non-ASN.1 seed encoding:
+ *
+ * - Bare seed (just the 32 bytes)
+ * - Bare priv (just the key bytes)
+ *
+ * A length of zero means that particular field is absent.
+ *
+ * The p8_shift is 0 when the top-level tag+length occupy four bytes, 2 when
+ * they occupy two by†es, and 4 when no tag is used at all.
+ */
+typedef struct {
```

Comment:
later down the track we can share this object

Reply:
Yes, when merged the `PKCS#8` format type and preference vectors are not algorithm specific and can be shared.